### PR TITLE
Fix gem update --user-install

### PR DESCRIPTION
**Pull request by bronzdoc**
# Description:

Closes https://github.com/rubygems/rubygems/issues/2886

Update only user insatalled gems when the `--user-install` flag is passed to `gem update`

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by duckinator** [comment]
@bronzdoc should this perhaps print a message with a quick explanation that non-user gems are being skipped + how to update those?

Other than that, this looks good to me overall, but would definitely be a breaking change so I'm not sure when it should be merged.

**Comment by bronzdoc** [comment]
@duckinator  I don't  know 🤔  `gem update --user-install` updating system-wide gems seems like a bug to me.

**Comment by MSP-Greg** [comment]
I agree with both of you.  But, I'd prefer the change.  I'm always updating my local ruby-head build, so all my gems used for testing are in --user-install...

**Comment by duckinator** [comment]
@bronzdoc thinking on it a bit more, I can see arguments in either direction. I'm cool with it being merged and treated as a bugfix if other folks are. :+1:

**Comment by deivid-rodriguez** [comment]
I'll review this tomorrow, :+1:. I do agree in principle with considering this a bug.

File: test/rubygems/test_gem_commands_update_command.rb [hunk 1]
```diff
@@ -373,7 +373,13 @@ def test_execute_user_install
     installer = @cmd.installer
     user_install = installer.instance_variable_get :@user_install
 
-    assert user_install, 'user_install must be set on the installer'
+    assert user_install, "user_install must be set on the installer"
```
Comment (deivid-rodriguez) [inline test/rubygems/test_gem_commands_update_command.rb]: Usual pickiness of mine of not changing unnecessary code just for style.

File: lib/rubygems/commands/update_command.rb [hunk 2]
```diff
@@ -137,6 +136,9 @@ def fetch_remote_gems(spec) # :nodoc:
   def highest_installed_gems # :nodoc:
     hig = {} # highest installed gems
 
+    # Get only gem specifications installed as --user-install
+    Gem::Specification.dirs = Gem.user_dir if options[:user_install]
```
Comment (deivid-rodriguez) [inline lib/rubygems/commands/update_command.rb]: I think this is the actual fix needed and none of the changes in `lib/rubygems/specification.rb` are actually needed?

Comment (deivid-rodriguez) [reply lib/rubygems/commands/update_command.rb]: I don't know where the new methods are being used? :thinking: I see what you're trying to do, but I don't think you finished the implementation? You might need to loop only through the newly added `Specification.user_installed_stubs` in the line below this one.

Also, we would need a new test to make sure that default gems are not updated.

Comment (bronzdoc) [reply lib/rubygems/commands/update_command.rb]: Hmm 🤔 you are right, I mixed things up.

I'll move the changes of `specification.rb` to another/followup PR

File: test/rubygems/test_gem_commands_update_command.rb (hunk 1 above)
Comment (bronzdoc) [reply test/rubygems/test_gem_commands_update_command.rb]: 🤔 I usually don't, unless is related  code that I'm modifying any way 🙈

**Comment by bronzdoc** [comment]
@bundlerbot try

**Comment by bronzdoc** [comment]
Build is not running after a force push 🙄

File: test/rubygems/test_gem_commands_update_command.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply test/rubygems/test_gem_commands_update_command.rb]: Right. Although I prefer to never touch style unless explicitly normalizing it, I'm also aware of that approach of "fix style as you go".

I won't bug you with this kind of comments any more if the code you're changing contains minor style tweaks next to it :+1:.

File: lib/rubygems/commands/update_command.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply lib/rubygems/commands/update_command.rb]: Yeah. This doesn't completely fix the bug, but it improves things over the current situation in my opinion, so let's do it. I'll open a separate ticket for the remaining issue.

**Comment by deivid-rodriguez** [comment]
Let's see how our bot is feeling now :)

@bundlerbot merge

**Pull request merged**